datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -70,320 +57,185 @@ impl Debug for InListExpr {
+    })
 }
 
-// try evaluate all list exprs and check if the exprs are constants or not
-fn try_cast_static_filter_to_set(
+fn evaluate_list(
     list: &[Arc<dyn PhysicalExpr>],
-    schema: &Schema,
-) -> Result<HashSet<ScalarValue>> {
-    let batch = RecordBatch::new_empty(Arc::new(schema.to_owned()));
-    list.iter()
-        .map(|expr| match expr.evaluate(&batch) {
-            Ok(ColumnarValue::Array(_)) => Err(DataFusionError::NotImplemented(
-                "InList doesn't support to evaluate the array result".to_string(),
-            )),
-            Ok(ColumnarValue::Scalar(s)) => Ok(s),
-            Err(e) => Err(e),
-        })
-        .collect::<Result<HashSet<_>>>()
-}
-
-fn make_list_contains_decimal(
-    array: &Decimal128Array,
-    list: Vec<ColumnarValue>,
-    negated: bool,
-) -> ColumnarValue {
-    let contains_null = list
-        .iter()
-        .any(|v| matches!(v, ColumnarValue::Scalar(s) if s.is_null()));
-    let values = list
+    batch: &RecordBatch,
+) -> Result<ArrayRef> {
+    let scalars = list
         .iter()
-        .flat_map(|v| match v {
-            ColumnarValue::Scalar(s) => match s {
-                Decimal128(v128op, _, _) => *v128op,
-                datatype => unreachable!(
-                    "InList can't reach other data type {} for {}.",
-                    datatype, s
-                ),
-            },
-            ColumnarValue::Array(_) => {
-                unimplemented!("InList does not yet support nested columns.")
-            }
-        })
-        .collect::<Vec<_>>();
-
-    collection_contains_check_decimal!(array, values, negated, contains_null)
-}
-
-fn make_set_contains_decimal(
-    array: &Decimal128Array,
-    set: &HashSet<ScalarValue>,
-    negated: bool,
-) -> ColumnarValue {
-    let contains_null = set.iter().any(|v| v.is_null());
-    let native_set = set
-        .iter()
-        .flat_map(|v| match v {
-            Decimal128(v128op, _, _) => *v128op,
-            datatype => {
-                unreachable!("InList can't reach other data type {} for {}.", datatype, v)
-            }
-        })
-        .collect::<HashSet<_>>();
-
-    collection_contains_check_decimal!(array, native_set, negated, contains_null)
-}
-
-fn set_contains_utf8<OffsetSize: OffsetSizeTrait>(
-    array: &GenericStringArray<OffsetSize>,
-    set: &HashSet<ScalarValue>,
-    negated: bool,
-) -> ColumnarValue {
-    let contains_null = set.iter().any(|v| v.is_null());
-    let native_set = set
-        .iter()
-        .flat_map(|v| match v {
-            Utf8(v) | LargeUtf8(v) => v.as_deref(),
-            datatype => {
-                unreachable!("InList can't reach other data type {} for {}.", datatype, v)
-            }
+        .map(|expr| {
+            expr.evaluate(batch).and_then(|r| match r {
+                ColumnarValue::Array(_) => Err(DataFusionError::Execution(
+                    "InList expression must evaluate to a scalar".to_string(),
```

Comment:
This is consistent with the code before, which instead returned "InList does not yet support nested columns". I think this error message is more clear about what the cause actually is, it has nothing to do with nested columns at all.

See #3766